File: LeetCode/1-100/2. Add Two Numbers/Solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *result = NULL, *temp = NULL, *prev = NULL, *i = l1, *j = l2;
        int carry = 0;
        while(i || j){
            carry += (i ? i->val : 0) + (j ? j->val : 0);
            temp = new struct ListNode(carry%10);
            if(result == NULL)
                result = temp;
            else
                prev->next = temp;
            prev = temp;
            carry /= 10;
            
            if(i != NULL)
                i = i->next;
            if(j != NULL)
                j = j->next;
        }
        if(carry > 0)
            temp->next = new struct ListNode(carry);
        
        return result;
    }
};

File: LeetCode/1-100/2. Add Two Numbers/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode prev1 = l1;
        ListNode curr1 = l1;
        ListNode curr2 = l2;
        int carry = 0;
        while (curr1 != null) {
            if (curr2 != null) {
                carry += curr2.val;
                curr2 = curr2.next;
            }
            carry += curr1.val;
            curr1.val = carry % 10;
            carry /= 10;
            prev1 = curr1;
            curr1 = curr1.next;
        }
        while (curr2 != null) {
            carry += curr2.val;
            prev1.next = new ListNode();
            prev1 = prev1.next;
            prev1.val = carry % 10;
            carry /= 10;
            curr2 = curr2.next;
        }
        if (carry != 0) {
            prev1.next = new ListNode(carry);
        }
        return l1;
    }
}